### PR TITLE
Functional tests for system components: compilers, hip, mpi.

### PR DESCRIPTION
**Pull request by mszpindler**
To be tested when upgraded LUMI is available again.

### CODE REVIEW

**Comment by mszpindler** [comment]
There are basically three elements here: simple compiler checks for actual version and openmp offload test (for gpu), hipinfo test querying for available gpus and osu benchmarks. I have added rocm support in osu checks and enabled validation. The validation always fails on gpus with device-to-device mode with unknown reason.

**Comment by mszpindler** [comment]
Also excluded alltoall tests because of very long run time.

**Comment by mszpindler** [comment]
For the `vAdd_ompGPU` test `LUMI/22.08` is needed.

**Comment by mszpindler** [comment]
> There are basically three elements here: simple compiler checks for actual version and openmp offload test (for gpu), hipinfo test querying for available gpus and osu benchmarks. I have added rocm support in osu checks and enabled validation. The validation always fails on gpus with device-to-device mode with unknown reason.

OSU Release notes say:
* Bug Fixes (since v6.0)
    - Add support for data validation for ROCm benchmarks
        * Thanks to Amir Shehata @ORNL for the report and initial patch

**Comment by mszpindler** [comment]
OSU now work with LUMI/22.08

File: system-tests/hip/hipInfo.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    executable = 'hipInfo'
+    build_locally = False
+    maintainers = ['mszpindler']
+    num_gpus_per_node = 8
+
+    @run_before('compile')
+    def set_compiler_flags(self):
+        self.build_system.cc = 'hipcc'
+        self.build_system.cxx = 'hipcc'
+        self.gpu_build = 'hip'
```
Comment (rsarm) [inline system-tests/hip/hipInfo.py:21]: This is not used, right?

File: system-tests/hip/hipInfo.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    sourcepath = 'hipInfo.cpp'
+    executable = 'hipInfo'
+    build_locally = False
+    maintainers = ['mszpindler']
+    num_gpus_per_node = 8
+
+    @run_before('compile')
+    def set_compiler_flags(self):
+        self.build_system.cc = 'hipcc'
+        self.build_system.cxx = 'hipcc'
```
Comment (rsarm) [inline system-tests/hip/hipInfo.py:20]: The compiler definition maybe we can move it to the `builtin-hip` definition in the configuration?

File: system-tests/mpi/osu/osu_check.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        },
+    }
+
+    @run_after('init')
+    def setup_by_scale(self):
+        with contextlib.suppress(KeyError):
+            self.reference = self.allref[self.num_nodes]
+
+    @run_after('init')
+    def setup_per_build_type(self):
```
Comment (rsarm) [inline system-tests/mpi/osu/osu_check.py:163]: Maybe most of this function can be moved to `lumi_osu_benchmarks` and be used both in the pt2pt and collective tests, each one then addeding the particular variables needed in each case?

File: system-tests/mpi/osu/osu_check.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        return sn.assert_found(rf'^{self.message_size}.*Pass', self.stdout)
+
+
+@rfm.simple_test
+class lumi_osu_collective_check(lumi_osu_benchmarks):
+    benchmark_info = parameter([
+        #('mpi.collective.osu_alltoall', 'latency'),
+        ('mpi.collective.osu_allreduce', 'latency'),
+    ], fmt=lambda x: x[0], loggable=True)
+    num_nodes = parameter([1, 2])
```
Comment (rsarm) [inline system-tests/mpi/osu/osu_check.py:137]: What about running it for a larger number of nodes? 8 o 16 maybe?

File: system-tests/hip/hipInfo.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+@rfm.simple_test
+class HipInfo(rfm.RegressionTest):
+    valid_systems = ['lumi:gpu']
+    valid_prog_environs = ['builtin-hip']
+    modules = ['rocm']
+    build_system = 'SingleSource'
+    sourcepath = 'hipInfo.cpp'
+    executable = 'hipInfo'
+    build_locally = False
```
Comment (rsarm) [inline system-tests/hip/hipInfo.py:13]: Does this need to be built on a compute node?

File: system-tests/mpi/osu/osu_check.py (hunk 4 above)
Comment (mszpindler) [reply system-tests/mpi/osu/osu_check.py:137]: 4 is max for `small` partition, I am staying with the limit for now

**Comment by mszpindler** [comment]
Comments resolved

**Pull request merged**